### PR TITLE
Construct FL course when server does not have data

### LINKED ISSUE #208: Support run FL courses when the server doesn't have data
In the current version, we need the `server_data` for building a model and getting a trainer, but in some cases, the server doesn't have the data or only has test data. 
Furthermore, does the server really need a trainer?  I find that the server's trainer is only used for evaluation.

### PR DESCRIPTION
**Pull request by xieyxclack**
As the title says. When the server does not have data:
- if some data is available as data representative (e.g., in simulation the client data), these data can be used to extract the input shape for building models at the server;
- Otherwise, the users can specify `model.input_shape` to tell what's the input shape (a tuple) to get the model if necessary;
- fix #208 

When the server does not have data in simulation, users can set `data.type=''`. (Please see the examples in scripts/run_distributed_lr.sh)

Please @rayrayraykk check `get_shape_from_data` in federatescope/core/auxiliaries/model_builder.py, thx!

### CODE REVIEW

File: federatedscope/core/auxiliaries/data_builder.py [hunk 1]
```diff
@@ -572,14 +572,20 @@ def get_data(config):
...
+        modified_config = config
     else:
         raise ValueError('Data {} not found.'.format(config.data.type))
 
     if config.federate.mode.lower() == 'standalone':
         return data, modified_config
     else:
         # Invalid data_idx
-        if config.distribute.data_idx not in data.keys():
+        if config.distribute.data_idx == -1:
```
Comment (rayrayraykk) [inline federatedscope/core/auxiliaries/data_builder.py:586]: When `data.type` is `None` or `""` while `distribute.data_idx` is given, we should assert in  `configs`.

File: federatedscope/core/configs/cfg_fl_setting.py [hunk 2]
```diff
@@ -49,7 +49,12 @@ def extend_fl_setting_cfg(cfg):
     cfg.distribute.client_port = 50050
     cfg.distribute.role = 'client'
     cfg.distribute.data_file = 'data'
```
Comment (rayrayraykk) [inline federatedscope/core/configs/cfg_fl_setting.py:51]: The data-related keyword `distribute.data_file` in distributed mode seems to be the same meaning as `data.root`.

**Review by rayrayraykk — COMMENTED** [review]
LGTM, but please see the inline comments, thanks.

File: federatedscope/core/auxiliaries/model_builder.py [hunk 3]
```diff
@@ -11,99 +11,138 @@
...
+        1): {train/val/test: {x:ndarray, y:ndarray}}}
+        2): {train/val/test: DataLoader}
+    Returns:
+        shape (tuple): the input shape
+    """
+    # Handle some special cases
+    if model_config.type.lower() in ['vmfnet', 'hmfnet']:
+        return data['train'].n_col if model_config.type.lower(
+        ) == 'vmfnet' else data['train'].n_row
+    elif model_config.type.lower() in [
```
Comment (rayrayraykk) [inline federatedscope/core/auxiliaries/model_builder.py:30]: For graph-level tasks, data is  `{train/val/test: DataLoader}`.

File: federatedscope/core/auxiliaries/data_builder.py (hunk 1 above)
Comment (xieyxclack) [reply federatedscope/core/auxiliaries/data_builder.py:586]: When `date.type` is None or "", the `distribute.data_idx` does not work. IMO, we can provide a WARNING but not an assertion here.

File: federatedscope/core/configs/cfg_fl_setting.py (hunk 2 above)
Comment (xieyxclack) [reply federatedscope/core/configs/cfg_fl_setting.py:51]: ‘data.root’ is a file directory and `distribute.data_file` includes the filename here

File: federatedscope/core/auxiliaries/model_builder.py (hunk 3 above)
Comment (xieyxclack) [reply federatedscope/core/auxiliaries/model_builder.py:30]: I have fixed this issue, thx!

**Review by yxdyc — APPROVED** [review]
Good Job. LGTM.

**Pull request merged**